cpp/src/parquet/chunker_internal.cc

```diff
@@ -0,0 +1,413 @@
+      // chunking process since the gearhash doesn't need to be updated
+      return;
+    }
+    rolling_hash_ = (rolling_hash_ << 1) + kGearhashTable[nth_run_][value];
+    has_matched_ = has_matched_ || ((rolling_hash_ & rolling_hash_mask_) == 0);
+  }
+
+  template <int ByteWidth>
+  void Roll(const uint8_t* value) {
+    // Update the rolling hash with a compile-time known sized value, set has_matched_ to
+    // true if the hash matches the mask.
+
+    chunk_size_ += ByteWidth;
+    if (chunk_size_ < min_chunk_size_) {
+      // short-circuit if we haven't reached the minimum chunk size, this speeds up the
+      // chunking process since the gearhash doesn't need to be updated
+      return;
+    }
+    for (size_t i = 0; i < ByteWidth; ++i) {
+      rolling_hash_ = (rolling_hash_ << 1) + kGearhashTable[nth_run_][value[i]];
+      has_matched_ = has_matched_ || ((rolling_hash_ & rolling_hash_mask_) == 0);
+    }
+  }
+
+  template <typename T>
+  void Roll(const T* value) {
+    return Roll<sizeof(T)>(reinterpret_cast<const uint8_t*>(value));
+  }
+
+  void Roll(const uint8_t* value, int64_t length) {
+    // Update the rolling hash with a binary-like value, set has_matched_ to true if the
+    // hash matches the mask.
+
+    chunk_size_ += length;
+    if (chunk_size_ < min_chunk_size_) {
+      // short-circuit if we haven't reached the minimum chunk size, this speeds up the
+      // chunking process since the gearhash doesn't need to be updated
+      return;
+    }
+    for (auto i = 0; i < length; ++i) {
+      rolling_hash_ = (rolling_hash_ << 1) + kGearhashTable[nth_run_][value[i]];
+      has_matched_ = has_matched_ || ((rolling_hash_ & rolling_hash_mask_) == 0);
+    }
+  }
+
+  bool NeedNewChunk() {
+    // decide whether to create a new chunk based on the rolling hash; has_matched_ is
+    // set to true if we encountered a match since the last NeedNewChunk() call
+    if (ARROW_PREDICT_FALSE(has_matched_)) {
+      has_matched_ = false;
+      // in order to have a normal distribution of chunk sizes, we only create a new chunk
+      // if the adjused mask matches the rolling hash 8 times in a row, each run uses a
+      // different gearhash table (gearhash's chunk size has geometric distribution, and
+      // we use central limit theorem to approximate normal distribution, see
+      // section 6.2.1 in paper https://www.cidrdb.org/cidr2023/papers/p43-low.pdf)
+      if (ARROW_PREDICT_FALSE(++nth_run_ >= 7)) {
+        nth_run_ = 0;
+        chunk_size_ = 0;
+        return true;
+      }
+    }
+    if (ARROW_PREDICT_FALSE(chunk_size_ >= max_chunk_size_)) {
+      // we have a hard limit on the maximum chunk size, note that we don't reset the
+      // rolling hash state here, so the next NeedNewChunk() call will continue from the
+      // current state
+      chunk_size_ = 0;
+      return true;
+    }
+    return false;
+  }
+
+  void ValidateChunks(const std::vector<Chunk>& chunks, int64_t num_levels) {
```

Comment:
Do we need `#ifndef NDEBUG` here?

```suggestion
#ifndef NDEBUG
  void ValidateChunks(const std::vector<Chunk>& chunks, int64_t num_levels) {
```

Reply:
Although `ValidateChunks` uses debug checks, it still has a loop which would be compiled in release builds. I am sparing that loop with the `NDEBUG` check to not call `ValidateChunks()` at all.